File: api/tts.py
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import asyncio
import edge_tts

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Access-Control-Allow-Origin', 'https://itm-kaiwa.github.io')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_GET(self):
        parsed_path = urlparse(self.path)
        query = parse_qs(parsed_path.query)

        text = query.get('text', [''])[0]
        voice = query.get('voice', ['ja-JP-NanamiNeural'])[0]

        if not text:
            self.send_response(400)
            self.send_header('Access-Control-Allow-Origin', 'https://itm-kaiwa.github.io')
            self.end_headers()
            self.wfile.write(b"Missing text parameter")
            return

        try:
            audio_data = asyncio.run(self.synthesize(text, voice))
            
            self.send_response(200)
            self.send_header('Content-Type', 'audio/mpeg')
            self.send_header('Access-Control-Allow-Origin', 'https://itm-kaiwa.github.io')
            self.send_header('Cache-Control', 'no-store')
            self.end_headers()
            self.wfile.write(audio_data)
        except Exception as e:
            self.send_response(500)
            self.send_header('Access-Control-Allow-Origin', 'https://itm-kaiwa.github.io')
            self.end_headers()
            self.wfile.write(f"TTS Error: {str(e)}".encode('utf-8'))

    async def synthesize(self, text, voice):
        communicate = edge_tts.Communicate(text, voice)
        audio_data = b""
        async for chunk in communicate.stream():
            if chunk["type"] == "audio":
                audio_data += chunk["data"]
        return audio_data

File: api/tts.js
const { MsEdgeTTS, OUTPUT_FORMAT } = require('msedge-tts');

const ALLOWED_ORIGIN = 'https://itm-kaiwa.github.io';

module.exports = async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', ALLOWED_ORIGIN);
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(204).end();
  }

  const text  = req.query.text;
  const voice = req.query.voice || 'ja-JP-NanamiNeural';

  if (!text) {
    return res.status(400).send('Missing text parameter');
  }

  console.log(`TTS request: voice=${voice}, text="${text}"`);

  try {
    const tts = new MsEdgeTTS();
    await tts.setMetadata(voice, OUTPUT_FORMAT.AUDIO_24KHZ_48KBITRATE_MONO_MP3);

    const { audioStream } = await tts.toStream(text);

    // Buffer all chunks before sending (more reliable in serverless)
    const chunks = [];
    for await (const chunk of audioStream) {
      chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
    }

    if (chunks.length === 0) {
      return res.status(500).send('TTS Error: No audio data received from Microsoft');
    }

    const audioBuffer = Buffer.concat(chunks);
    console.log(`TTS success: ${audioBuffer.length} bytes`);

    res.setHeader('Content-Type', 'audio/mpeg');
    res.setHeader('Content-Length', audioBuffer.length);
    res.setHeader('Cache-Control', 'no-store');
    return res.status(200).send(audioBuffer);

  } catch (err) {
    const message = err?.message || err?.toString() || 'Unknown error';
    console.error('TTS error:', message, err);
    return res.status(500).send(`TTS Error: ${message}`);
  }
};
